### PR TITLE
Allow to define only setter in computed property

### PR DESCRIPTION
**Pull request by Exelord**
This PR will fix issue: #15939 

Before when we tried to access undefined getter we got into this problem:
screen shot 2017-12-09 at 11 32 24

### CODE REVIEW

File: packages/ember-metal/lib/computed.js
Comment (bekzod) [inline packages/ember-metal/lib/computed.js]: I think it will be cleaner to put empty function outside of constructor, like `function K(){}` as it is done in other places, maybe it should also contain `warning` maybe that user hasn't passed getter this should be confirmed with core devs I guess

Comment (Exelord) [reply packages/ember-metal/lib/computed.js]: Sure. I fix the problem in the source now. Could you take a look?
All remarks have been mentioned in the issue. Definitely worth to discuss.

File: packages/ember-metal/lib/computed.js
Comment (bekzod) [inline packages/ember-metal/lib/computed.js]: I don't think it is best solution, since `this._getter` won't be provided in super narrow case, and in common case it is provided, the best solution is to provide empty function as you did earlier

**Review by bekzod — CHANGES_REQUESTED** [review]

File: packages/ember-metal/lib/computed.js
Comment (bekzod) [reply packages/ember-metal/lib/computed.js]: `noop` function for `_getter` in constructor is best solution I think

Comment (Exelord) [reply packages/ember-metal/lib/computed.js]: Well, it depends. If we use the `noop` function in the constructor we will allocate a memory even if we never call the `get`.

Right, now the solution is more complex but it fixes the source of the problem and saves resources. I noticed that when I look at `set` function definition. Now, they are implemented in a similar way.

Do you think this is a bad approach?

Comment (bekzod) [reply packages/ember-metal/lib/computed.js]: > Well, it depends. If we use the noop function in the constructor we will allocate a memory even if we never call the get.

thats why I asked to keep `noop` function outside constructor ([example](https://github.com/emberjs/ember.js/blob/master/packages/ember-glimmer/lib/renderer.ts#L184)), also not sure if it should contain warning inside `warn('getter missing .....')` this should be confirmed by core team I guess

> Right, now the solution is more complex but it fixes the source of the problem and saves resources. I noticed that when I looked at set function definition. Now, they are implemented in a similar way.

Not only it is complex, it is also adds extra overhead/processing to be done each time even `getter` is provided (which is 99% of time)

Thats why your first approach was best

Comment (bekzod) [reply packages/ember-metal/lib/computed.js]: it is hotpath code therefore slightly more work will result in big overheads

Comment (Exelord) [reply packages/ember-metal/lib/computed.js]: You suggested: `noop function for _getter in constructor is best solution I think` not, outside :p

Ok, will try that one. But isn't Ember.K already deprecated? Shouldn't we just use `() {}`?

Comment (bekzod) [reply packages/ember-metal/lib/computed.js]: > You suggested: noop function for _getter in constructor is best solution I think not, outside :p
https://github.com/emberjs/ember.js/pull/15962#discussion_r155918993

> Ok, will try that one. But isn't Ember.K already deprecated? Shouldn't we just use () {}? 

global `Ember.K` yes deprecated, but using `noop` function  naming `K` is not :P, but calling it `noop` is more descriptive I think

**Review by bekzod — APPROVED** [review]

**Comment by rwjblue** [comment]
@Exelord - Thanks for working on this!  I totally agree that we should address the issue you raised in https://github.com/emberjs/ember.js/issues/15939, but I don't think this is the right path forward. Instead of making a setter only CP silently return `undefined`, we should make it error (essentially emulating the behavior of `.readOnly`).

Something like:

```js
ComputedPropertyPrototype.get = function(obj, keyName) {
  assert(`Cannot call '.get' to retrieve ${keyName} on ${obj}, it has to getter.', typeof this._getter === 'function');
  // ...snip...
};
```

What do you think?

**Comment by Exelord** [comment]
Thanks, @rwjblue :)
So, I see here 2 issues:

## 1. Overriding setter
### Problem
You are not allowed to override a setter and have a default behavior of getter without defining it.
```js
  currentUser: computed({
    get() {}, // need to be defined 😞 | right now, it will return the set cached value as expected

    set(key, value) {
      this.get('notifier').configure({ payload: { person: value } });
      return value;
    }
  }),
```

### Solution
I think the current state of the PR is perfectly fixing that issue. Right now you can just have a default getter behavior and override only a setter to extend his role.

## 2. There is no possibility to have a setter only property
This issue is mentioned by you. I agree that we should raise an error in case when we are trying to reach a getter analogically like the `readOnly` does with the setter, but for this purpose, I would like to create sth like `writeOnly` cp which will cover that case exclusively. Otherwise, we will kill the first possibility to just extend the setter responsibility. What do you think?

**Comment by rwjblue** [comment]
> You are not allowed to override a setter and have a default behavior of getter without defining it.

Can you unpack this a bit? What does this mean? If you try to define a setter only CP it throws an error? When?

**Comment by Exelord** [comment]
Basically, if you would have sth like:
```js
currentUser: computed({
    set(key, value) {
      this.get('notifier').configure({ payload: { person: value } });
      return value;
    }
  }),
```
and set the value `this.set('currentUser', { name: 'me' })` everything will work but then if you would try to get it by `this.get('currentUser')` it will raise an error (the missing get function which wasn't predicted by the implementation). That's what this PR fixes. (test's scenario)

So, right now you can't override a setter without defining an empty getter because you will fall into that issue.

**Comment by rwjblue** [comment]
> So, right now you can't override a setter without defining an empty getter because you will fall into that issue when you will try to use the getter on it.

Yes, this makes sense to me. There is no getter to call, this isn’t something we should silently normalize it is an actual error in the users program. Also, If you override a CP you need to override both portions (if you don’t need to customize the getter, you can have its implementation be `return this._super()`).

Having a setter only CP (in general) seems “fine”, but if someone tries to `.get` it then it should fire a helpful assertion stating that the computed properly has no getter.

To reiterate: 

1) we should support `.set` for a setter only CP
2) attempting to `.get` a setter only CP should provide a useful error (the current error is horribly unhelpful)

**Comment by Exelord** [comment]
Hmm... I'm not talking here about set-only properties. I'm talking about extending the setter. Analogically to getter I can omit the setter to just define custom getter but I can not omit the getter to define custom setter:
```js
currentUser: computed({
    get() {
      return 'test';
    }
}),
```
In that case, the cp is not get-only, it just has a custom getter. The same case should allow defining custom setter.

> Also, If you override a CP you need to override both portions

According to the docs and the code `set` and `get` are not the required keys, they are just allowed and can occur also in `or` relation.

> (if you don’t need to customize the getter, you can have its implementation returns this._super())

The solution here is just to define an empty function - it returns undefined which means that I will use a cache for it (the default behavior). The `super` shouldn't work here as there isn't exist any parent of `get`. get is not defined by default which causes the problem. In contrast, to set if the set is not defined there is fired a custom behaviour with caching.

**Comment by bekzod** [comment]
I think best solution is to have `noop` function with warning inside, 
```javascript
function getNoop(val) {
  warn('You are missing getter, and you called get .....')
  return val
}

//then in constructor 
 this._getter = config.get || getNoop;
 this._setter = config.set

```

**Comment by Exelord** [comment]
If we add the warning here we should also warn when there is only getter passed. So basically we should require defining them both.

**Comment by bekzod** [comment]
> ...should also warn when there is only getter passed. 

passing only `setter` is unusual thats why warning is needed, passing only `getter` is not unusual most of the time only `getter` passed since it is computed property :)

**Comment by Exelord** [comment]
Overriding setters can be useful when would you like to avoid using observers. (not always applicable).
Maybe the case is not usual but it against the documentation and assertions:
> Config object passed to computed can only contain `get` or `set` keys.
> Computed properties must receive a getter or a setter, you passed none.

From the logical point of view the current behaviour is arguing the reality.

**Comment by rwjblue** [comment]
I don’t see a conflict with the documentation at all. If you only need a getter: All works fine. If you only need a setter: All works fine. The problem here is that you are only passing a setter but then somehow expect the getter to have some behavior too.

**Comment by Exelord** [comment]
If you pass only a getter you can still use a working setter, but if you pass only a setter you can't use a working getter.

> Config object passed to computed can only contain get or set keys.

That sentence means that you can pass `get` or `set` (in real, you can pass `get` AND `set`) but indeed you can pass [`get` OR (`get` AND `set`)]
which is arguing with: 
> Computed properties must receive a getter or a setter, you passed none.

`get` || `set` !== `get`|| (`get` && `set`)

**Comment by Serabe** [comment]
IMHO, set only properties should return `undefined` when they are `get`ted.

In vanilla JS:

```js
class MyClass {
  set hola(value) {
    console.log(value);
  }
}

let obj = new MyClass();
obj.hola = "hi!"; // outputs "hi!"
console.log(obj.hola); // outputs undefined
```

As we are trying to get closer, and closer to plain JS behaviour, having `obj.get('hola')` differ from `obj.hola` in JS classes seems wrong to me.

**Comment by Exelord** [comment]
@rwjblue any more thoughts on that? Is it still not relevant? :/

**Comment by mmun** [comment]
I agree with @Serabe.

**Comment by mmun** [comment]
I'm persuaded by the argument that it is more likely for a developer to accidentally crash the Ember Inspector by clicking on such a property, or for something like QUnit to accidentally throw an error when its trying to serialize the assertion message, than it is for a developer to accidentally `get` a `set`-only property in their application code.

I'm happy to merge this as is after we squash all the commits down.

**Comment by mmun** [comment]
@rwjblue what do you think of @bekzod's idea to have a development-only warning? Seems like a win-win.

File: packages/ember-metal/lib/computed.js
Comment (bekzod) [inline packages/ember-metal/lib/computed.js]: maybe /* noop */ comment not needed

**Comment by rwjblue** [comment]
Seems like a good compromise....

**Comment by Exelord** [comment]
```js
class MyClass {
  set hola(value) {
    console.log(value);
  }
}
```

The native getter doesn't fire any warning if it's not defined.

For me, it's natural that I want to override only a setter and keep a default getter behavior. 
Set only property doesn't mean it's write-only as get only property doesn't mean is read-only.

**Comment by mmun** [comment]
On second thought, the warning would be almost as annoying in the inspector / JSON.stringify case. Happy to merge as is once the tests pass.

**Pull request merged**
